### ISSUE TITLE
False Positive `dict-init-mutate` when dict assignment references dict

### ISSUE THREAD

**Issue by clavedeluna** [PyCQA/pylint#7997]
### Bug description

Initially brought up here https://github.com/PyCQA/pylint/pull/7794#issuecomment-1366726866

```
expected_counts = {"int64": 1, "object": 1}
expected_counts[dtype_name] = 3 + expected_counts.get(dtype_name, 0)
```
from (pandas](https://github.com/pandas-dev/pandas/blob/d1ecf63e2040daecad6e0a8485a35e8a23393795/pandas/tests/reshape/test_get_dummies.py#L110)

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
-init-mutate:
Declare all known key/values when initializing the dictionary.
```

### Expected behavior

no dict-init-mutate msg

### Pylint version

```shell
latest
```

**Comment by Pierre-Sassoulas** [comment]
Isn't the corrected code the following ?
```python
key = "something"
counts = {"int64": 1, "object": 1, key: 3}
```


We already know that the default of 0 will have to be used because key is not defined yet.

Now if the base code is:
```python
key = "something"
counts = {"int64": 1, "object": 1, key: 5}
counts[key] = 3 + counts.get(key, 0)
```

Then the corrected code can also be changed to:
```python
key = "something"
counts = {"int64": 1, "object": 1, key: 3 + 5}
```

**Comment by nickdrozd** [comment]
The more ambitious way to address this would be to recursively check the assign value to ensure that the dictionary name doesn't appear anywhere. The less ambitious way would be to only emit the warning if the value is simple (say, constants).

Either way this is kind of a yucky pattern, precisely because it can't be inlined easily.

**Comment by uzzell** [comment]
In the Pandas example, the key `dtype_name` is not a constant, but (is derived from) a function argument.  So, we do not know _a priori_ what value needs to be changed.  The following case is similar to but simpler than the Pandas example:
```
def make_dict(key):
    d = {'a': 1, 'b': 2}
    d[key] = 3 + d.get(key, 0)
```

**Comment by clavedeluna** [comment]
Yes @uzzell good point, I tried to provide a simplified example of the issue but by doing so I simplified it too much and it seemed trivial. I think both cases of `key` being a func arg and a var should still be considered FP.
